### PR TITLE
updateUriRecFeedback API in Curator & Shoebox + Unit test

### PR DESCRIPTION
**Pull request by lawzlo**
@stkem @yingjieMiao @leogrim @andrewconner

### CODE REVIEW

File: kifi-backend/modules/common/app/com/keepit/model/UriRecommendationFeedback.scala [hunk 1]
```diff
@@ -0,0 +1,15 @@
+package com.keepit.model
+
+import com.keepit.common.util.EnumFormatUtil
+import play.api.libs.json.{ Writes, Reads }
+
+object UriRecommendationFeedback extends Enumeration {
```
Comment (stkem) [inline kifi-backend/modules/common/app/com/keepit/model/UriRecommendationFeedback.scala]: I think that since this has only a few fields and is not going to grow arbitrarily a case class would be the better choise to store it instead of an enumeration. That way type safety is fully preserved.

Comment (lawzlo) [inline kifi-backend/modules/common/app/com/keepit/model/UriRecommendationFeedback.scala]: Yes. This might not for grow. Just for avoiding string in the code for easy maintenance. I think case class does same thing. But enumeration has string name with value, so it's easy to parse from/to json. I guess?

Comment (stkem) [inline kifi-backend/modules/common/app/com/keepit/model/UriRecommendationFeedback.scala]: True, but case classes are easy to serialize too, thanks to macros (e.g. `@json`) and you avoid the `Map[String, _]` which is not type safe.

File: kifi-backend/modules/shoebox/test/com/keepit/controllers/website/RecommendationsControllerTest.scala [hunk 2]
```diff
@@ -1,28 +1,75 @@
...
+import play.api.libs.json.{ Json }
+import play.api.mvc.SimpleResult
+import play.api.test.Helpers._
+import play.api.test.{ FakeRequest }
 
 import scala.concurrent.ExecutionContext.Implicits.global
+import scala.concurrent.Future
 
-class RecommendationsControllerTest extends Specification with ShoeboxTestInjector {
+class RecommendationsControllerTest extends Specification with ShoeboxApplicationInjector {
```
Comment (lawzlo) [inline kifi-backend/modules/shoebox/test/com/keepit/controllers/website/RecommendationsControllerTest.scala]: Tried not using running application but not working.

Comment (stkem) [inline kifi-backend/modules/shoebox/test/com/keepit/controllers/website/RecommendationsControllerTest.scala]: Check out this blog post on how to test a controller without the application: http://eng.kifi.com/speed-up-play-tests-direct/
A good example is the LibraryControllerTest

File: kifi-backend/modules/common/app/com/keepit/model/UriRecommendationFeedback.scala (hunk 1 above)
Comment (lawzlo) [inline kifi-backend/modules/common/app/com/keepit/model/UriRecommendationFeedback.scala]: I'm changing to use case class. Actually I think you are right. It looks cleaner.

File: kifi-backend/modules/shoebox/test/com/keepit/controllers/website/RecommendationsControllerTest.scala (hunk 2 above)
Comment (lawzlo) [inline kifi-backend/modules/shoebox/test/com/keepit/controllers/website/RecommendationsControllerTest.scala]: `LibraryControllerTest` is the example I looked at, I also looked at `HealthcheckModuleTest` which the one I worked on fixing day. Still not getting it work. :(

Comment (lawzlo) [inline kifi-backend/modules/shoebox/test/com/keepit/controllers/website/RecommendationsControllerTest.scala]: A ah, I removed `FakeShoeboxAppSecureSocialModule()` and it works!

File: kifi-backend/modules/common/app/com/keepit/model/UriRecommendationFeedback.scala [hunk 3]
```diff
@@ -0,0 +1,10 @@
+package com.keepit.model
+
+import com.kifi.macros.json
+
+@json case class UriRecommendationFeedback(
+    val seen: Option[Boolean],
+    val clicked: Option[Boolean],
+    val kept: Option[Boolean]) {
+  override def toString = s"seen:$seen --- clicked:$clicked --- kept:$kept"
```
Comment (stkem) [inline kifi-backend/modules/common/app/com/keepit/model/UriRecommendationFeedback.scala]: I would throw the type in here, makes for better error messages, i.e. `s"UriRecommendationFeeback(..."`

File: kifi-backend/modules/common/test/com/keepit/curator/FakeCuratorServiceClientImpl.scala [hunk 4]
```diff
@@ -9,13 +7,13 @@ import com.keepit.common.service.ServiceType
...
+import com.keepit.model.{ UriRecommendationFeedback, ScoreType, NormalizedURI, User }
 import com.keepit.curator.model.RecommendationInfo
 
 class FakeCuratorServiceClientImpl(val airbrakeNotifier: AirbrakeNotifier) extends CuratorServiceClient {
   val serviceCluster: ServiceCluster = new ServiceCluster(ServiceType.TEST_MODE, Providers.of(airbrakeNotifier), new FakeScheduler(), () => {})
   protected def httpClient: com.keepit.common.net.HttpClient = ???
 
-  def adHocRecos(userId: Id[User], n: Int, scoreCoefficientsUpdate: Map[ScoreType, Float]): Future[Seq[RecommendationInfo]] = Future.successful(Seq.empty)
-
+  def adHocRecos(userId: Id[User], n: Int, scoreCoefficientsUpdate: Map[ScoreType.Value, Float]): Future[Seq[RecommendationInfo]] = Future.successful(Seq.empty)
```
Comment (stkem) [inline kifi-backend/modules/common/test/com/keepit/curator/FakeCuratorServiceClientImpl.scala]: Why this change? This appears to be working correctly as it is right now.

Comment (lawzlo) [inline kifi-backend/modules/common/test/com/keepit/curator/FakeCuratorServiceClientImpl.scala]: I think we didn't use this in Prod code since it's FakeCuratorServiceClientImpl. ScoreType.Value should be the right one I think.

File: kifi-backend/modules/common/app/com/keepit/model/UriRecommendationFeedback.scala (hunk 3 above)
Comment (lawzlo) [inline kifi-backend/modules/common/app/com/keepit/model/UriRecommendationFeedback.scala]: Will fix it. Thanks.

File: kifi-backend/modules/curator/app/com/keepit/curator/model/UriRecommendationRepo.scala [hunk 5]
```diff
@@ -45,6 +46,10 @@ class UriRecommendationRepoImpl @Inject() (
   def table(tag: Tag) = new RecommendationTable(tag)
   initTable()
 
+  override def save(uriRecommendation: UriRecommendation)(implicit session: RWSession): UriRecommendation = {
```
Comment (stkem) [inline kifi-backend/modules/curator/app/com/keepit/curator/model/UriRecommendationRepo.scala]: This override doesn't seem to be doing anything extra.

Comment (lawzlo) [inline kifi-backend/modules/curator/app/com/keepit/curator/model/UriRecommendationRepo.scala]: Deleting it.

File: kifi-backend/modules/curator/app/com/keepit/curator/model/UriRecommendationRepo.scala [hunk 6]
```diff
@@ -53,6 +58,11 @@ class UriRecommendationRepoImpl @Inject() (
     (for (row <- rows if row.userId === userId) yield row).sortBy(_.masterScore.desc).take(maxBatchSize).list
   }
 
+  def updateUriRecommendationFeedback(userId: Id[User], uriId: Id[NormalizedURI], feedback: UriRecommendationFeedback)(implicit session: RSession): Boolean = {
+    val row = (for (row <- rows if row.uriId === uriId && row.userId === userId) yield row)
+    row.update(row.firstOption.get.withUpdateFeedback(feedback)) > 0
```
Comment (stkem) [inline kifi-backend/modules/curator/app/com/keepit/curator/model/UriRecommendationRepo.scala]: This invites race conditions with the actual recommendation generation (i.e. if that row is being generated at the same time). Please use a direct update query on only the fields changed (instead of reading the object and writing it back)

File: kifi-backend/modules/curator/app/com/keepit/curator/controllers/internal/CuratorController.scala [hunk 7]
```diff
@@ -23,4 +23,15 @@ class CuratorController @Inject() (recoGenCommander: RecommendationGenerationCom
...
   }
 
+  def updateUriRecommendationFeedback(userId: Id[User], uriId: Id[NormalizedURI]) = Action.async { request =>
+    val json = request.body.asJson
+    json match {
+      case Some(json) => {
+        val feedback = json.as[UriRecommendationFeedback]
+        recoGenCommander.updateUriRecommendationFeedback(userId, uriId, feedback).map(update => Ok(Json.toJson(update)))
+      }
+      case None => Future.successful(Ok(Json.toJson(false)))
```
Comment (stkem) [inline kifi-backend/modules/curator/app/com/keepit/curator/controllers/internal/CuratorController.scala]: This is an internal call, thus you can expect the data coming in to be in the correct format => invalid json should be a hard exception.

File: kifi-backend/modules/common/test/com/keepit/curator/FakeCuratorServiceClientImpl.scala (hunk 4 above)
Comment (stkem) [inline kifi-backend/modules/common/test/com/keepit/curator/FakeCuratorServiceClientImpl.scala]: Ah, I see. And that is why string keyed maps can be problematic :-) 
(Still the right solution in this case though, I think)

File: kifi-backend/modules/curator/app/com/keepit/curator/model/UriRecommendationRepo.scala (hunk 6 above)
Comment (lawzlo) [inline kifi-backend/modules/curator/app/com/keepit/curator/model/UriRecommendationRepo.scala]: Good point. Fixing it. Thanks.

File: kifi-backend/modules/curator/app/com/keepit/curator/controllers/internal/CuratorController.scala (hunk 7 above)
Comment (lawzlo) [inline kifi-backend/modules/curator/app/com/keepit/curator/controllers/internal/CuratorController.scala]: I agree. But I still like error handling in every controller layer. In case some bad things happen in the future. (Someone change external call controller or network issue)

File: kifi-backend/modules/common/test/com/keepit/curator/FakeCuratorServiceClientImpl.scala (hunk 4 above)
Comment (lawzlo) [inline kifi-backend/modules/common/test/com/keepit/curator/FakeCuratorServiceClientImpl.scala]: Changing it.

File: kifi-backend/modules/curator/app/com/keepit/curator/controllers/internal/CuratorController.scala (hunk 7 above)
Comment (stkem) [inline kifi-backend/modules/curator/app/com/keepit/curator/controllers/internal/CuratorController.scala]: Yes, and this way it will gracefully return when instead we want to know that there was an error. Throwing an exception (by simple assuming the json to be correct and letting it fail) will trigger an airbrake message.

Comment (lawzlo) [inline kifi-backend/modules/curator/app/com/keepit/curator/controllers/internal/CuratorController.scala]: Adding throw exception

File: kifi-backend/modules/curator/app/com/keepit/curator/controllers/internal/CuratorController.scala [hunk 8]
```diff
@@ -14,13 +11,26 @@ import play.api.libs.concurrent.Execution.Implicits.defaultContext
...
   }
 
+  def updateUriRecommendationFeedback(userId: Id[User], uriId: Id[NormalizedURI]) = Action.async { request =>
+    val json = request.body.asJson
+    json match {
+      case Some(json) => {
+        val feedback = json.as[UriRecommendationFeedback]
+        recoGenCommander.updateUriRecommendationFeedback(userId, uriId, feedback).map(update => Ok(Json.toJson(update)))
+      }
+      case None => Future.failed(new IllegalArgumentException("request body should have feedback body"))
```
Comment (stkem) [inline kifi-backend/modules/curator/app/com/keepit/curator/controllers/internal/CuratorController.scala]: Typically what we do is simply get the value out of the body (i.e. `request.body.asJson.get`) or use Plays transparent tolerant json parser (a good example is the `kifiHit` method in ShoeboxController.scala)
That way we get a good error message and stack trace for free.

Comment (lawzlo) [inline kifi-backend/modules/curator/app/com/keepit/curator/controllers/internal/CuratorController.scala]: Ok. I will use `get` to get value from body.

File: kifi-backend/modules/common/app/com/keepit/common/util/MapFormatUtil.scala [hunk 9]
```diff
@@ -8,8 +7,9 @@ import com.keepit.common.json._
 
 object MapFormatUtil {
 
-  implicit def scoreTypeMapFormat(implicit f: Format[Map[String, Float]]): Format[Map[ScoreType.Value, Float]] =
-    mapKeyFormat(ScoreType.withName _, _.toString)
+  //  Example of how to convert customized key to json when using a map
+  //  implicit def scoreTypeMapFormat(implicit f: Format[Map[String, Float]]): Format[Map[ScoreType.Value, Float]] =
+  //    mapKeyFormat(ScoreType.withName _, _.toString)
```
Comment (stkem) [inline kifi-backend/modules/common/app/com/keepit/common/util/MapFormatUtil.scala]: Commented out code left. Is this intended to be used again really soon?

Comment (lawzlo) [inline kifi-backend/modules/common/app/com/keepit/common/util/MapFormatUtil.scala]: Deleting it.

File: kifi-backend/modules/common/app/com/keepit/model/UriRecommendationScores.scala [hunk 10]
```diff
@@ -0,0 +1,12 @@
+package com.keepit.model
+import com.kifi.macros.json
+
+@json case class UriRecommendationScores(
```
Comment (stkem) [inline kifi-backend/modules/common/app/com/keepit/model/UriRecommendationScores.scala:4]: Oh, how come you changed this one to a case class as well?

File: kifi-backend/modules/curator/app/com/keepit/curator/model/UriRecommendationRepo.scala [hunk 11]
```diff
@@ -53,6 +54,12 @@ class UriRecommendationRepoImpl @Inject() (
     (for (row <- rows if row.userId === userId) yield row).sortBy(_.masterScore.desc).take(maxBatchSize).list
   }
 
+  def updateUriRecommendationFeedback(userId: Id[User], uriId: Id[NormalizedURI], feedback: UriRecommendationFeedback)(implicit session: RSession): Boolean = {
```
Comment (stkem) [inline kifi-backend/modules/curator/app/com/keepit/curator/model/UriRecommendationRepo.scala:57]: forgot to mention this earlier, but all update queries should also update the `updated_at` column to `currentDateTime`

Comment (lawzlo) [inline kifi-backend/modules/curator/app/com/keepit/curator/model/UriRecommendationRepo.scala:57]: cool. will change it.

File: kifi-backend/modules/shoebox/app/com/keepit/controllers/website/RecommendationsController.scala [hunk 12]
```diff
@@ -19,15 +19,17 @@ class RecommendationsController @Inject() (
...
 
-  def adHocRecos(n: Int) = JsonAction.authenticatedAsync { request =>
+  def adHocRecos(n: Int) = JsonAction.authenticatedParseJsonAsync { request =>
     if (userExperimentCommander.userHasExperiment(request.userId, ExperimentType.ADMIN)) {
-      val body = request.body.asJson match {
-        case Some(json) => json.as[Map[ScoreType.Value, Float]]
-        case None => Map[ScoreType.Value, Float]()
-      }
-      commander.adHocRecos(request.userId, n, body).map(fkis => Ok(Json.toJson(fkis)))
+      val scores = request.body.as[UriRecommendationScores]
```
Comment (stkem) [inline kifi-backend/modules/shoebox/app/com/keepit/controllers/website/RecommendationsController.scala:24]: I'm not sure how the macro based json serializer behaves with missing fields (which we need to support here). Can you test this somehow?

**Comment by stkem** [comment]
made a few more comments. In particular the missing fields on deserialization one is important. lgtm otherwise

File: kifi-backend/modules/shoebox/app/com/keepit/controllers/website/RecommendationsController.scala (hunk 12 above)
Comment (lawzlo) [inline kifi-backend/modules/shoebox/app/com/keepit/controllers/website/RecommendationsController.scala:24]: Changing unit test a little bit(missing a field). No issue with missing field.

File: kifi-backend/modules/common/app/com/keepit/model/UriRecommendationScores.scala (hunk 10 above)
Comment (lawzlo) [inline kifi-backend/modules/common/app/com/keepit/model/UriRecommendationScores.scala:4]: Same way with RecFeedback one :)

**Comment by lawzlo** [comment]
Compiled, tested. Pushing to Prod.
